components/aws/sagemaker/common/_utils.py

```diff
@@ -113,17 +122,49 @@ def print_logs_for_job(cw_client, log_grp, job_name):
         logging.error(e)
 
 
-def get_sagemaker_client(region, endpoint_url=None):
+def get_boto3_session(region, role_arn=None):
+    """Creates a boto3 session, optionally assuming a role"""
+
+    # By default return a basic session
+    if role_arn is None:
+        return Session()
+
+    def refresh():
+        client = boto3.client("sts", region_name=region,
+                                endpoint_url='https://sts.' + region + '.amazonaws.com')
+        response = client.assume_role(RoleArn=role_arn, RoleSessionName=ASSUME_SESSION_NAME_FORMAT).get("Credentials")
+        return dict(
+            access_key=response.get("AccessKeyId"),
+            secret_key=response.get("SecretAccessKey"),
+            token=response.get("SessionToken"),
+            expiry_time=response.get("Expiration").isoformat()
+        )
+
+    session_credentials = RefreshableCredentials.create_from_metadata(
+        metadata=refresh(),
+        refresh_using=refresh,
+        method='sts-assume-role'
+    )
+
+    current_session = get_session()
+    # Must modify this internally, no clean way to set this otherwise
+    current_session._credentials = session_credentials
+    current_session.set_config_variable('region', region)
+    return Session(botocore_session=current_session)
+
+def get_sagemaker_client(region, endpoint_url=None, assume_role_arn=None):
     """Builds a client to the AWS SageMaker API."""
-    session_config = botocore.config.Config(
+    session = get_boto3_session(region, assume_role_arn)
+    session_config = Config(
         user_agent='sagemaker-on-kubeflow-pipelines-v{}'.format(get_component_version())
     )
-    client = boto3.client('sagemaker', region_name=region, endpoint_url=endpoint_url, config=session_config)
+    client = session.client('sagemaker', region_name=region, endpoint_url=endpoint_url, config=session_config)
```

Comment:
will setting config here override the existing config returned from get_boto3_session's session object? if so, we need to merge both

Reply:
Good point. Will look into it. Either way it's easy to merge them.

Reply:
From reading [the Session documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html) there is no conflict here. Specifying this config attempts to override any properties on the lower level version, but since we're only overriding the user-agent this shouldn't affect it. Manual testing confirms this as well.

[More docs for Config](https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html)

Reply:
lets merge it so there is only one place to set it. Can be a separate method

Reply:
What do you mean merge it? This is the only place where `user_agent` is set.